File: .scripts/inventory_docs.js
#!/usr/bin/env node
/**
 * Documentation Inventory Script for Surrentumlabs (Node.js Version)
 * 
 * This script recursively scans /home/cinder/Documents/repos/Surrentumlabs/** 
 * for markdown documentation files and generates a JSON cache with metadata.
 * 
 * Author: Created by Agent Mode for Surrentumlabs documentation management
 * Purpose: Inventory all documentation files and metadata across Surrentumlabs
 * Created: 2024-12-22
 * 
 * Target files:
 * - README.md, AGENTS.md, CHANGELOG.md, plan.md, toaster.md
 * - docs/README.md, *.md inside docs/ directory
 * 
 * For each file captures:
 * - absolute path, repository root, repo name, repo category,
 * - primary language(s), last-modified timestamp (git log -1 --format=%ct FILE)
 * 
 * Outputs to: .scripts/cache/docs_index.json
 */

const fs = require('fs').promises;
const path = require('path');
const { execSync, spawn } = require('child_process');
const { promisify } = require('util');

// Configuration
const BASE_PATH = '/home/cinder/Documents/repos/Surrentumlabs';
const OUTPUT_DIR = '.scripts/cache';
const OUTPUT_FILE = path.join(OUTPUT_DIR, 'docs_index_node.json');

const TARGET_FILES = new Set([
    'README.md',
    'AGENTS.md', 
    'CHANGELOG.md',
    'plan.md',
    'toaster.md'
]);

const DOCS_PATTERNS = [
    'docs/README.md',
    'docs/*.md'
];

/**
 * Get git last modified timestamp for a file
 */
async function getGitLastModified(filePath) {
    try {
        const repoDir = path.dirname(filePath);
        const timestamp = execSync(
            `git log -1 --format=%ct "${filePath}"`,
            { cwd: repoDir, encoding: 'utf8' }
        ).trim();
        return parseInt(timestamp) || null;
    } catch (error) {
        // Fallback to file system timestamp
        try {
            const stats = await fs.stat(filePath);
            return Math.floor(stats.mtime.getTime() / 1000);
        } catch (fallbackError) {
            return null;
        }
    }
}

/**
 * Determine repository category based on parent directory
 */
function determineRepoCategory(repoPath) {
    const parentDir = path.basename(path.dirname(repoPath));
    
    const categoryMapping = {
        'Monorepos': 'Monorepo',
        'Rust Component Repos': 'Rust Component',
        'Typescript Component Repos': 'TypeScript Component',
        'Other': 'Other',
        'whiply_project': 'Other'
    };
    
    return categoryMapping[parentDir] || 'Other';
}

/**
 * Detect primary programming languages in a repository
 */
async function detectPrimaryLanguages(repoPath) {
    const languages = new Set();
    
    const languageIndicators = {
        'Rust': ['Cargo.toml', 'Cargo.lock', 'src/main.rs', 'src/lib.rs'],
        'TypeScript': ['package.json', 'tsconfig.json'],
        'JavaScript': ['package.json'],
        'Python': ['requirements.txt', 'setup.py', 'pyproject.toml'],
        'Go': ['go.mod', 'go.sum'],
        'Java': ['pom.xml', 'build.gradle'],
        'C++': ['CMakeLists.txt', 'Makefile'],
        'C': ['Makefile']
    };
    
    for (const [lang, indicators] of Object.entries(languageIndicators)) {
        for (const indicator of indicators) {
            if (indicator.includes('*')) {
                // Handle glob patterns
                const extension = indicator.replace('*', '');
                try {
                    const files = await fs.readdir(repoPath);
                    if (files.some(file => file.endsWith(extension))) {
                        languages.add(lang);
                        break;
                    }
                } catch (error) {
                    // Ignore read errors
                }
            } else {
                // Handle exact file matches
                try {
                    await fs.access(path.join(repoPath, indicator));
                    languages.add(lang);
                    break;
                } catch (error) {
                    // File doesn't exist, continue
                }
            }
        }
    }
    
    // Special handling for TypeScript/JavaScript
    if (languages.has('TypeScript')) {
        languages.add('JavaScript');
    }
    
    // If no languages detected, default to Markdown
    if (languages.size === 0) {
        languages.add('Markdown');
    }
    
    return Array.from(languages);
}

/**
 * Find repository root by looking for .git directory
 */
async function findRepositoryRoot(filePath) {
    let currentPath = path.dirname(filePath);
    
    while (currentPath !== '/') {
        try {
            await fs.access(path.join(currentPath, '.git'));
            return currentPath;
        } catch (error) {
            currentPath = path.dirname(currentPath);
        }
    }
    
    // If no .git found, assume the directory of the file is repo root
    return path.dirname(filePath);
}

/**
 * Recursively walk directory and find matching files
 */
async function* walkDirectory(dir) {
    try {
        const entries = await fs.readdir(dir, { withFileTypes: true });
        
        for (const entry of entries) {
            const fullPath = path.join(dir, entry.name);
            
            if (entry.isDirectory()) {
                // Skip .git directories
                if (entry.name !== '.git') {
                    yield* walkDirectory(fullPath);
                }
            } else if (entry.isFile() && entry.name.endsWith('.md')) {
                yield fullPath;
            }
        }
    } catch (error) {
        // Ignore permission errors and continue
        console.error(`Error accessing ${dir}: ${error.message}`);
    }
}

/**
 * Check if file should be included based on target patterns
 */
function shouldIncludeFile(filePath) {
    const filename = path.basename(filePath);
    
    // Check exact filename matches
    if (TARGET_FILES.has(filename)) {
        return true;
    }
    
    // Check docs directory patterns
    if (filePath.includes('/docs/') && filename.endsWith('.md')) {
        return true;
    }
    
    return false;
}

/**
 * Process a single file and extract metadata
 */
async function processFile(filePath) {
    try {
        const repoRoot = await findRepositoryRoot(filePath);
        const repoName = path.basename(repoRoot);
        const repoCategory = determineRepoCategory(repoRoot);
        const primaryLanguages = await detectPrimaryLanguages(repoRoot);
        const lastModified = await getGitLastModified(filePath);
        const relativePath = path.relative(repoRoot, filePath);
        
        const docInfo = {
            absolute_path: filePath,
            repository_root: repoRoot,
            repo_name: repoName,
            repo_category: repoCategory,
            primary_languages: primaryLanguages,
            last_modified_timestamp: lastModified,
            last_modified_date: lastModified ? new Date(lastModified * 1000).toISOString() : null,
            filename: path.basename(filePath),
            relative_path: relativePath
        };
        
        console.log(`Found: ${filePath}`);
        return docInfo;
    } catch (error) {
        console.error(`Error processing ${filePath}: ${error.message}`);
        return null;
    }
}

/**
 * Generate summary statistics
 */
function generateStatistics(docsInventory) {
    const totalFiles = docsInventory.length;
    const repositoriesFound = new Set(docsInventory.map(doc => doc.repo_name)).size;
    
    const categories = {};
    const languages = {};
    
    for (const doc of docsInventory) {
        // Count by category
        categories[doc.repo_category] = (categories[doc.repo_category] || 0) + 1;
        
        // Count by languages
        for (const lang of doc.primary_languages) {
            languages[lang] = (languages[lang] || 0) + 1;
        }
    }
    
    return {
        totalFiles,
        repositoriesFound,
        categories,
        languages
    };
}

/**
 * Main function
 */
async function main() {
    console.log('Starting documentation inventory scan (Node.js version)');
    console.log(`Base path: ${BASE_PATH}`);
    console.log(`Target files: ${Array.from(TARGET_FILES).join(', ')}`);
    console.log('Target docs patterns: docs/README.md, docs/*.md');
    console.log('-'.repeat(60));
    
    // Check if base path exists
    try {
        await fs.access(BASE_PATH);
    } catch (error) {
        console.error(`ERROR: Base path does not exist: ${BASE_PATH}`);
        process.exit(1);
    }
    
    // Ensure output directory exists
    await fs.mkdir(OUTPUT_DIR, { recursive: true });
    
    const docsInventory = [];
    
    // Process all markdown files
    for await (const filePath of walkDirectory(BASE_PATH)) {
        if (shouldIncludeFile(filePath)) {
            const docInfo = await processFile(filePath);
            if (docInfo) {
                docsInventory.push(docInfo);
            }
        }
    }
    
    // Sort by repository name and filename
    docsInventory.sort((a, b) => {
        const repoCompare = a.repo_name.localeCompare(b.repo_name);
        return repoCompare !== 0 ? repoCompare : a.filename.localeCompare(b.filename);
    });
    
    // Generate statistics
    const stats = generateStatistics(docsInventory);
    
    // Create output structure
    const output = {
        metadata: {
            scan_timestamp: new Date().toISOString(),
            base_path: BASE_PATH,
            total_files: stats.totalFiles,
            repositories_found: stats.repositoriesFound,
            categories: stats.categories,
            languages: stats.languages,
            target_files: Array.from(TARGET_FILES),
            docs_patterns: DOCS_PATTERNS
        },
        documentation_files: docsInventory
    };
    
    // Write JSON output
    await fs.writeFile(OUTPUT_FILE, JSON.stringify(output, null, 2), 'utf8');
    
    console.log('-'.repeat(60));
    console.log('Scan complete (Node.js version)!');
    console.log(`Total files found: ${stats.totalFiles}`);
    console.log(`Repositories scanned: ${stats.repositoriesFound}`);
    console.log(`Categories: ${Object.entries(stats.categories).map(([k, v]) => `${k}(${v})`).join(', ')}`);
    console.log(`Primary languages: ${Object.entries(stats.languages).map(([k, v]) => `${k}(${v})`).join(', ')}`);
    console.log(`Results written to: ${OUTPUT_FILE}`);
}

// Run the main function
if (require.main === module) {
    main().catch(error => {
        console.error('Error:', error);
        process.exit(1);
    });
}

module.exports = {
    main,
    processFile,
    detectPrimaryLanguages,
    findRepositoryRoot,
    determineRepoCategory
};

File: .scripts/inventory_docs.py
#!/usr/bin/env python3
"""
Documentation Inventory Script for Surrentumlabs

This script recursively scans /home/cinder/Documents/repos/Surrentumlabs/** for 
markdown documentation files and generates a JSON cache with metadata.

Author: Created by Agent Mode for Surrentumlabs documentation management
Purpose: Inventory all documentation files and metadata across Surrentumlabs
Created: 2024-12-22

Target files:
- README.md
- AGENTS.md  
- CHANGELOG.md
- plan.md
- toaster.md
- docs/README.md
- *.md inside docs/ directory

For each file captures:
- absolute path
- repository root
- repo name
- repo category (Monorepo, Rust Component, TypeScript Component, Other)
- primary language(s)
- last-modified timestamp (git log -1 --format=%ct FILE)

Outputs to: .scripts/cache/docs_index.json
"""

import os
import json
import subprocess
import sys
from pathlib import Path
from typing import Dict, List, Optional, Set
from datetime import datetime

# Base path to scan
BASE_PATH = "/home/cinder/Documents/repos/Surrentumlabs"

# Target filenames and patterns
TARGET_FILES = {
    "README.md",
    "AGENTS.md", 
    "CHANGELOG.md",
    "plan.md",
    "toaster.md"
}

# Documentation patterns
DOCS_PATTERNS = {
    "docs/README.md",
    "docs/*.md"
}

def get_git_last_modified(file_path: str) -> Optional[int]:
    """
    Get the last modified timestamp using git log -1 --format=%ct FILE
    Returns Unix timestamp or None if git command fails
    """
    try:
        result = subprocess.run(
            ["git", "log", "-1", "--format=%ct", file_path],
            capture_output=True,
            text=True,
            cwd=os.path.dirname(file_path)
        )
        if result.returncode == 0 and result.stdout.strip():
            return int(result.stdout.strip())
    except (subprocess.SubprocessError, ValueError):
        pass
    
    # Fallback to file system modification time
    try:
        stat_info = os.stat(file_path)
        return int(stat_info.st_mtime)
    except OSError:
        return None

def determine_repo_category(repo_path: str) -> str:
    """
    Determine repository category based on its parent directory
    """
    parent_dir = os.path.basename(os.path.dirname(repo_path))
    
    category_mapping = {
        "Monorepos": "Monorepo",
        "Rust Component Repos": "Rust Component", 
        "Typescript Component Repos": "TypeScript Component",
        "Other": "Other",
        "whiply_project": "Other"
    }
    
    return category_mapping.get(parent_dir, "Other")

def detect_primary_languages(repo_path: str) -> List[str]:
    """
    Detect primary programming languages in a repository
    """
    languages = []
    
    # Check for common language indicators
    language_indicators = {
        "Rust": ["Cargo.toml", "Cargo.lock", "src/main.rs", "src/lib.rs"],
        "TypeScript": ["package.json", "tsconfig.json", "*.ts", "*.tsx"],
        "JavaScript": ["package.json", "*.js", "*.jsx"],
        "Python": ["requirements.txt", "setup.py", "pyproject.toml", "*.py"],
        "Go": ["go.mod", "go.sum", "*.go"],
        "Java": ["pom.xml", "build.gradle", "*.java"],
        "C++": ["CMakeLists.txt", "Makefile", "*.cpp", "*.cc", "*.cxx"],
        "C": ["Makefile", "*.c", "*.h"]
    }
    
    for lang, indicators in language_indicators.items():
        for indicator in indicators:
            if "*" in indicator:
                # Handle glob patterns
                pattern = indicator.replace("*", "")
                if any(f.endswith(pattern) for f in os.listdir(repo_path) if os.path.isfile(os.path.join(repo_path, f))):
                    languages.append(lang)
                    break
            else:
                # Handle exact file matches
                if os.path.exists(os.path.join(repo_path, indicator)):
                    languages.append(lang)
                    break
    
    # If no languages detected, try to infer from directory structure
    if not languages:
        try:
            files = os.listdir(repo_path)
            if any(f.endswith('.md') for f in files):
                languages.append("Markdown")
        except OSError:
            pass
    
    return languages if languages else ["Unknown"]

def find_repository_root(file_path: str) -> str:
    """
    Find the repository root by looking for .git directory
    """
    current_path = os.path.dirname(file_path)
    
    while current_path != "/":
        if os.path.exists(os.path.join(current_path, ".git")):
            return current_path
        current_path = os.path.dirname(current_path)
    
    # If no .git found, assume the immediate parent of the file is repo root
    return os.path.dirname(file_path)

def get_repo_name(repo_root: str) -> str:
    """
    Extract repository name from the repository root path
    """
    return os.path.basename(repo_root)

def scan_for_docs(base_path: str) -> List[Dict]:
    """
    Recursively scan for documentation files and collect metadata
    """
    docs_inventory = []
    
    for root, dirs, files in os.walk(base_path):
        # Skip .git directories
        dirs[:] = [d for d in dirs if d != '.git']
        
        for file in files:
            file_path = os.path.join(root, file)
            
            # Check if file matches our target patterns
            should_include = False
            
            # Check exact filename matches
            if file in TARGET_FILES:
                should_include = True
            
            # Check docs directory patterns
            if "docs" in root.split(os.sep):
                if file == "README.md" or file.endswith(".md"):
                    should_include = True
            
            if should_include:
                try:
                    repo_root = find_repository_root(file_path)
                    repo_name = get_repo_name(repo_root)
                    repo_category = determine_repo_category(repo_root)
                    primary_languages = detect_primary_languages(repo_root)
                    last_modified = get_git_last_modified(file_path)
                    
                    doc_info = {
                        "absolute_path": file_path,
                        "repository_root": repo_root,
                        "repo_name": repo_name,
                        "repo_category": repo_category,
                        "primary_languages": primary_languages,
                        "last_modified_timestamp": last_modified,
                        "last_modified_date": datetime.fromtimestamp(last_modified).isoformat() if last_modified else None,
                        "filename": file,
                        "relative_path": os.path.relpath(file_path, repo_root)
                    }
                    
                    docs_inventory.append(doc_info)
                    print(f"Found: {file_path}")
                    
                except Exception as e:
                    print(f"Error processing {file_path}: {e}", file=sys.stderr)
                    continue
    
    return docs_inventory

def main():
    """
    Main function to run the documentation inventory
    """
    print(f"Starting documentation inventory scan of: {BASE_PATH}")
    print(f"Target files: {', '.join(TARGET_FILES)}")
    print(f"Target docs patterns: docs/README.md, docs/*.md")
    print("-" * 60)
    
    if not os.path.exists(BASE_PATH):
        print(f"ERROR: Base path does not exist: {BASE_PATH}")
        sys.exit(1)
    
    # Scan for documentation files
    docs_inventory = scan_for_docs(BASE_PATH)
    
    # Sort by repository name and then by filename for consistent output
    docs_inventory.sort(key=lambda x: (x["repo_name"], x["filename"]))
    
    # Generate summary statistics
    total_files = len(docs_inventory)
    repos_found = len(set(doc["repo_name"] for doc in docs_inventory))
    categories = {}
    languages = {}
    
    for doc in docs_inventory:
        # Count by category
        category = doc["repo_category"]
        categories[category] = categories.get(category, 0) + 1
        
        # Count by languages
        for lang in doc["primary_languages"]:
            languages[lang] = languages.get(lang, 0) + 1
    
    # Create output structure
    output = {
        "metadata": {
            "scan_timestamp": datetime.now().isoformat(),
            "base_path": BASE_PATH,
            "total_files": total_files,
            "repositories_found": repos_found,
            "categories": categories,
            "languages": languages,
            "target_files": list(TARGET_FILES),
            "docs_patterns": list(DOCS_PATTERNS)
        },
        "documentation_files": docs_inventory
    }
    
    # Ensure output directory exists
    output_dir = ".scripts/cache"
    os.makedirs(output_dir, exist_ok=True)
    
    # Write JSON output
    output_file = os.path.join(output_dir, "docs_index.json")
    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(output, f, indent=2, ensure_ascii=False)
    
    print("-" * 60)
    print(f"Scan complete!")
    print(f"Total files found: {total_files}")
    print(f"Repositories scanned: {repos_found}")
    print(f"Categories: {', '.join(f'{k}({v})' for k, v in categories.items())}")
    print(f"Primary languages: {', '.join(f'{k}({v})' for k, v in languages.items())}")
    print(f"Results written to: {output_file}")

if __name__ == "__main__":
    main()
